pkg/api/secretgenerator/secretgenerator.go

```diff
@@ -0,0 +1,148 @@
+package secretgenerator
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/getlantern/deepcopy"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/util/gzip"
+)
+
+func LoadConfigFromPath(path string) (Config, error) {
+	var config Config
+	cfgBytes, err := gzip.ReadFileMaybeGZIP(path)
+	if err != nil {
+		return nil, err
+	}
+	if err := yaml.Unmarshal(cfgBytes, &config); err != nil {
+		return nil, err
+	}
+	return config, nil
+}
+
+type Config []SecretItem
+
+func (c *Config) UnmarshalJSON(data []byte) error {
+	var errs []error
+	var newConfig Config
+
+	var config []SecretItem
+	if err := json.Unmarshal(data, &config); err != nil {
+		return err
+	}
+
+	for _, si := range config {
+		items, err := si.generateItemsFromParams()
+		if err != nil {
+			errs = append(errs, err)
+		}
+		newConfig = append(newConfig, items...)
+	}
+	*c = newConfig
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (c Config) itemsByName() map[string][]SecretItem {
+	byName := make(map[string][]SecretItem)
+
+	for _, secretItem := range c {
+		byName[secretItem.ItemName] = append(byName[secretItem.ItemName], secretItem)
+	}
+
+	return byName
+}
+
+func (c Config) IsItemGenerated(name string) bool {
+	byName := c.itemsByName()
+	if _, ok := byName[name]; ok {
+		return true
+	}
+	return false
+}
+
+func (c Config) IsFieldGenerated(name, component string) bool {
+	byName := c.itemsByName()
+	if _, ok := byName[name]; !ok {
+		return false
+	}
+
+	for _, item := range byName[name] {
+		for _, field := range item.Fields {
+			if component == field.Name {
+				return true
+			}
+		}
+
+		for _, attachment := range item.Attachments {
+			if component == attachment.Name {
+				return true
+			}
+		}
+
+		if component == item.Password {
+			return true
+		}
+	}
+	return false
+}
+
+type FieldGenerator struct {
+	Name string `json:"name,omitempty"`
+	Cmd  string `json:"cmd,omitempty"`
+}
+
+type SecretItem struct {
+	ItemName    string              `json:"item_name"`
+	Fields      []FieldGenerator    `json:"fields,omitempty"`
+	Attachments []FieldGenerator    `json:"attachments,omitempty"`
+	Password    string              `json:"password,omitempty"`
+	Notes       string              `json:"notes,omitempty"`
+	Params      map[string][]string `json:"params,omitempty"`
+}
+
+func (si SecretItem) generateItemsFromParams() ([]SecretItem, error) {
+	var errs []error
+	var processedBwItems []SecretItem
+
+	replaceParameter := func(paramName, param, template string) string {
+		return strings.ReplaceAll(template, fmt.Sprintf("$(%s)", paramName), param)
+	}
+
+	itemsProcessingHolder := []SecretItem{si}
```

Comment:
```suggestion
	var result []SecretItem
```

Reply:
That change will break it. We need `si` to be in the slice in order to update it multiple times after the first combination will finish and the next will start.

Reply:
For the record, that code is a pretty much a copy-paste from https://github.com/openshift/ci-tools/blob/master/cmd/ci-secret-generator/main.go#L168-L203

Reply:
> That change will break it. We need si to be in the slice in order to update it multiple times after the first combination will finish and the next will start.

We don't, the only write access to it is `itemsProcessingHolder = itemsProcessed` which overwrites it entirely.

Reply:
Apparently, applying this change breaks the entire algorithm by returning an empty object. Either is something completely wrong here or I am missing something. For the record, I didn't change anything from our production code.

Reply:
Yeah, because we loop over the `itemsProcessingHolder` in the beginning instead of over `si`

Reply:
Yes, this seems intentional. I didn't touch anything. Let me check this out and refactor it from scratch.

Reply:
Also, we can't loop over `si` since it's not a slice, and we don't want to change it at all. When I first saw this code I thought that maybe it's better to re-write it from scratch but I didn't do it.